Feature/weeping willow workflow

erinstargazer opened the pull request:
This has the pipeline changes for the weeping willow merge in and a phylo profile module added. I'll add the other analysis modules to this PR or a new one depending on the timing.

erinstargazer committed:
Merging in weeping willow changes.

erinstargazer committed:
Some module clean up and fixes.

erinstargazer committed:
Adding analysis modules

erinstargazer committed:
Added zoogle table output, won't work until clinvar is added though.

erinstargazer committed:
Removed zoogle, hgnc, and aggregate phylo.

austinhpatton committed:
minimal famsa module

austinhpatton committed:
remove non-standard AAs

austinhpatton committed:
replace non-standard AAs with ambiguous

erinstargazer committed:
Removed old docker flag, and removed logging when phylo dist is skipping a gene family.

austinhpatton committed:
rename directory

austinhpatton committed:
more non-standard AA handling

erinstargazer committed:
Add in batch changes.

austinhpatton commented on subworkflows/local/infer_trees.nf:
```diff
@@ -0,0 +1,31 @@
+include { CIALIGN as TRIM_MSAS } from '../../modules/local/cialign'
+include { FASTTREE             } from '../../modules/local/fasttree'
+include { WITCH as ALIGN_SEQS  } from '../../modules/local/witch'
```
Hmmm.... I thought in one of my previous commits i overrode `WITCH as ALIGN_SEQS` with `FAMSA as ALIGN_SEQS`...

austinhpatton commented on nextflow_schema.json:
```diff
@@ -30,6 +30,14 @@
+                    "default": "full",
+                    "enum": ["full", "simplified"],
+                    "description": "Workflow mode: 'full' runs the complete pipeline with all optional analyses (BUSCO, IQTREE_PMSF, GENERAX_PER_FAMILY); 'simplified' runs a streamlined high-throughput workflow optimized for large datasets.",
+                    "help_text": "The 'full' mode (default) provides maximum flexibility and includes quality assessments (BUSCO) and detailed per-family GeneRax analyses. The 'simplified' mode uses a fixed toolchain (WITCH->CIAlign->FastTree) and skips optional steps for faster execution on large datasets.",
```
Just another note about the above comment regarding full/simplified/zoogle workflow modes.

austinhpatton commented on main.nf:
```diff
@@ -77,25 +81,44 @@ if (ch_tree_method == "iqtree") {
+    include { INFER_TREES as INFER_SPECIES_TREES    } from './subworkflows/local/infer_trees'
+    include { INFER_TREES as INFER_REMAINING_TREES  } from './subworkflows/local/infer_trees'
+}
+
```
Ahhhh, i see. That wasn't clicking with me until I saw this - so the simplified workflow necessarily hardcodes the methods used. So if we wanted a simplified and zoogle one, we'd just need to turn the last few modules for conservation core/analyses etc to a subworkflow and only do those steps when running in zoogle mode?

austinhpatton commented on main.nf:
```diff
@@ -105,14 +128,16 @@ include { ORTHOFINDER_PHYLOHOGS                     } from './modules/local/orth
+    include { BUSCO as BUSCO_BROAD                      } from './modules/nf-core-modified/busco'
+    include { DIAMOND_BLASTP as DIAMOND_BLASTP_TEST     } from './modules/nf-core-modified/diamond_blastp'
+    include { IQTREE_PMSF as IQTREE_PMSF_ALL            } from './modules/nf-core-modified/iqtree_pmsf'
+    include { IQTREE_PMSF as IQTREE_PMSF_REMAINING      } from './modules/nf-core-modified/iqtree_pmsf'
```
Honestly we should probably be just using the IQTREE module, rather than the IQTREE_PMSF (unless requested?) - the latter is significantly more involved than IQTREE, which already is much more heavy than FASTTREE

austinhpatton commented on main.nf:
```diff
@@ -305,106 +344,132 @@ workflow NOVELTREE {
+            ch_core_gene_trees = INFER_TREES.out.phylogeny
+            ch_rem_gene_trees = INFER_REMAINING_TREES.out.phylogeny
+            core_gene_tree_list = ch_core_gene_trees.collect { it[1] }
+        }
```
In effect this would mean that IQTREE would never be used unless the PMSF implementation is being requested, no? That doesn't seem like the intended functionality...

austinhpatton commented on main.nf:
```diff
@@ -450,19 +515,40 @@ workflow NOVELTREE {
+            ogs_ch,
+            ORTHOFINDER_MCL_ALL.out.inflation_dir
+        )
+    }
```
Honestly we probably always want to include the phylo profiles analyses into any/all modes, as these really is just reformatting of these results into more useful tabular formats. So, I think we can probably just take this out of the simplified mode.

austinhpatton commented on main.nf:
For some reason this version of main.nf in the pull request doesn't include all of the phylo dist steps yet. Seems like we need to make sure that's included here? I can see it on the branch though.

austinhpatton reviewed (changes requested):
So I've gone through and left some comments throughout. The biggest thing that stands out to me is that it seems like the phylo dist module didn't actually make it in to this PR, though it is on the branch? We'll definitely want to make sure that it is. 

Furthermore, it's probably best if we treat that as a subworkflow that's only invoked if running it in "zoogle" mode, which is like simplified mode, but just uses all the fastest options and tacks on the phylo dist module at the end. The simplified version can then use the fast implementations, skipping extraneous steps (including the phylo dist module), and then the full mode just does everything. 

What do you think? There's a smattering of other comments/questions/suggestions throughout but happy to tackle this together tomorrow.

austinhpatton commented on README.md:
````diff
@@ -17,7 +17,19 @@
+make docker-physicochemical-props
+make docker-phylo-dist
+```
+
````
Do we actually want to include these instructions? I guess maybe this is helpful/useful, especially for these more involved docker images?

austinhpatton commented on README.md:
```diff
@@ -35,6 +47,63 @@ Nextflow requires some memory resources to be allocated for overhead - consequen
+**Requirements:**
+- AWS Batch compute environment and job queue must be configured
+- Work directory and output directory must be S3 buckets
+- Appropriate IAM permissions for Batch and S3 access
```
Related to the above, probably worthwhile specifying that all of the workflow docker images also need to be built and hosted somewhere on a repository that can be accessed by the workflow?

austinhpatton commented on README.md:
```diff
@@ -35,6 +47,63 @@ Nextflow requires some memory resources to be allocated for overhead - consequen
+
+### Technical Details
+
+All Docker images are built from the repository root with the build context set to ensure access to vendored code in `bin/raas/`. The images are built for `linux/amd64` platform for compatibility.
```
```suggestion
All Docker images are built from the repository root with the build context set to ensure access to vendored code in `bin/phylo_dist/`. The images are built for `linux/amd64` platform for compatibility.
```

austinhpatton commented on bin/phylo_dist/README.md:
```diff
@@ -0,0 +1,80 @@
+- Noveltree includes ALL protein comparisons without filtering (no ClinVar integration)
+- Disease annotation columns (`disease_name`, etc.) are `NA` since pipeline uses `clinvar = NULL`
+
+**To add disease filtering:** Integrate ClinVar data by updating `modules/local/phylo_dist.nf` to pass `clinvar` parameter instead of NULL, then filter the aggregated output to `!is.na(disease_name)`.
```
Can we just get rid of this?
```suggestion
```

austinhpatton commented on docker/famsa/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+WORKDIR /data
+
+# Default command
+CMD ["/bin/bash", "-l"]
```
Newline @ EOF
```suggestion
CMD ["/bin/bash", "-l"]

```

austinhpatton commented on modules/local/famsa.nf:
```diff
@@ -0,0 +1,60 @@
+    def prefix = fasta.baseName
+    """
+    # Be sure to remove any non-standard amino acid codes in the input sequences, as this
+    # can cause errors downstream and in parsing.
```
```suggestion
    # Be sure to remove any non-standard amino acid codes in the input sequences, as this
    # can cause errors downstream and in parsing. Here, we replace with ambiguous AAs.
```

austinhpatton commented on main.nf:
```diff
@@ -532,23 +436,89 @@ workflow NOVELTREE {
+        ch_phylo_dist_input,
+        ch_speciesrax,
+        params.ref_species
+    )
```
I commented about this elsewhere, but I'm wondering if we should wrap this up into a subworkflow that only runs when the "workflow mode" is "zoogle" (or possibly also in "full"?) - I'm just thinking that since this is a non-trivial computational burden, it might only make sense to actually go through the process of running it if it's actually needed/requested? I'm open to opinions on this. 

```suggestion
    //
    // MODULE: PHYLO_DIST
    // Calculate phylogenetically-corrected protein distances
    //
    // Create channel pairing gene family trees with their physicochemical properties
    ch_phylo_dist_input = GENERAX_PER_SPECIES.out.generax_per_spp_gfts
        .map { meta, tree -> [meta, tree] }
        .combine(PHYSICOCHEMICAL_PROPS.out.per_family_summaries.flatten())
        .filter { meta, tree, props_file ->
            props_file.name.contains(meta.og) && props_file.name.contains("_summary_statistics.csv")
        }
        .filter { meta, tree, props_file ->
            // Validate gene family has sufficient proteins for phylo-dist analysis
            // Read CSV and extract protein IDs (first column, skip header)
            def lines = props_file.readLines()
            def proteinIds = lines.drop(1).collect { it.split(',')[0] }

            if (proteinIds.size() == 0) {
                log.info "Skipping ${meta.og}: No proteins found in CSV"
                return false
            }

            // Count reference species proteins
            def refCount = proteinIds.count { it.startsWith("${params.ref_species}_") }

            // Count non-reference proteins
            def nonrefCount = proteinIds.size() - refCount

            // Count proteins per non-reference species
            // NOTE: Must match R script's species extraction logic (line 145 of protein_distance_calculation_functions.R)
            // R uses: gsub("_.*", "", focal_prots) which removes everything after FIRST underscore
            def nonrefProteinsBySpecies = proteinIds
                .findAll { !it.startsWith("${params.ref_species}_") }
                .collect { it.replaceFirst(/_.*/, '') }  // Extract genus name only (everything before first underscore)
                .countBy { it }  // Map of species -> count

            // Count unique non-reference species
            def nonrefSpeciesCount = nonrefProteinsBySpecies.size()

            // Count how many non-reference species have at least 2 proteins
            // (Wilcoxon test requires at least 2 observations per group)
            def speciesWithEnoughProteins = nonrefProteinsBySpecies.count { species, count -> count >= 2 }

            // Apply validation criteria (maps directly to the 3 observed errors)
            def isValid = (refCount >= 1) && (nonrefCount >= 2) && (speciesWithEnoughProteins >= 2)

            return isValid
        }

    PHYLO_DIST(
        ch_phylo_dist_input,
        ch_speciesrax,
        params.ref_species
    )
```

austinhpatton reviewed (changes requested):
Okay, now that everything is here, my earlier comments regarding breaking out phylo_dist into its own subworkflow and other comments/suggestions still stand, but otherwise it all looks pretty good! Will push some new documentation for FAMSA, etc, soon. I also think that we can/should pare down some of the documentation that was added regarding the source of some of this code for the organismal prioritization - at the very least we should be sure we point to the repo associated with the pub, but I suggested some changes to that effect.

austinhpatton committed:
documentation updates

erinstargazer committed:
Cleaned up workflow to use subworkflow for all paths.

erinstargazer committed:
Fixed conflict.

erinstargazer committed:
Some PR cleanups and Makefile updated for all images.

erinstargazer committed:
Changes for singularity, and updated name for zoogle flag.

erinstargazer committed:
Very basic singularity doc added.

erinstargazer committed:
Fix zoogle profile.

erinstargazer committed:
Include lima settings.

erinstargazer committed:
Timetree updates.

austinhpatton commented on bin/phylo_dist/time_calibrate_species_tree.R:
```diff
@@ -0,0 +1,106 @@
+if (length(args) < 4) {
+  stop(paste(
+    "Usage: time_calibrate_species_tree.R <consensus_tree> <reference_tree> <output_tree> <calibration_method>",
+    "\n  consensus_tree: Path to consensus species tree (Newick format)",
```
I might suggest switching from "consensus tree" to "species tree" in any references/parameter/variable names for clarity, since this will always be the non-time-calibrated species tree, and "consensus" is a bit confusing.

austinhpatton commented on docker/phylo_dist/Dockerfile:
```diff
@@ -12,8 +12,39 @@ RUN apt-get update && \
+    gcc PATHd8.c -O3 -lm -o PATHd8 && \
+    mv PATHd8 /usr/local/bin/ && \
+    cd / && \
+    rm -rf /opt/PATHd8_src /tmp/PATHd8.zip
```
I'm glad this didn't cause any problems! Getting this to work was initially quite annoying...

austinhpatton reviewed (changes requested):
Tiny little naming change, but looks good!

erinstargazer committed:
Updated from using consensus tree workding, to species tree.

erinstargazer committed:
Don't need to update version for phylo_dist as it's a new docker image.

erinstargazer committed:
Update container name to be onsistent with rest of containers.

erinstargazer committed:
Be more specific about files being published to avoid race condition.

erinstargazer committed:
Fix generax config settings.

erinstargazer committed:
Increase memory for phylo profile generation to a high memory instance.

erinstargazer committed:
Try using chunks to help with memory usage.

erinstargazer committed:
Smaller chunk size and add debugging to ensure expected behavior.

erinstargazer committed:
Use batches.

austinhpatton committed:
updated documentation

austinhpatton committed:
updated rare AA handling

austinhpatton committed:
bump generax version

austinhpatton committed:
updated docker readme

austinhpatton committed:
consistent references

austinhpatton committed:
updated references

erinstargazer committed:
Add back in some fixes and cleanups that were lost.

erinstargazer committed:
The zoogle dataset doesn't use the same header format.

erinstargazer committed:
Handle zoogle header format in another place.

erinstargazer committed:
Force ultrametic.

erinstargazer committed:
Normalize species names: convert hyphens to underscores for consistent matching

erinstargazer committed:
Don't use collect when a channel is still needed.

erinstargazer committed:
For zoogle mode, use homo sapien for the ref species.

erinstargazer committed:
Fix physicochemical props to run on individual masa files.

erinstargazer committed:
Fix typo.

erinstargazer committed:
Keep timetree same, but update parsing so it's consistent with other modules.

erinstargazer committed:
Add newline at end of Dockerfile as suggested in PR feedback.

erinstargazer committed:
Attempt to add fasttree fallback.

When iqtree is used and fasttree fallback is enabled, this will
try iqtree and then fallback to fasttree if iqtree doesn't
generate a tree.

austinhpatton committed:
output restructuring, pmsf removal, and more

erinstargazer committed:
Fix link file name change handling. Remove old param no longer used.

austinhpatton committed:
speciesrax output cleanup

erinstargazer committed:
fix ref name when using human

austinhpatton committed:
always use treePL for congruification

erinstargazer committed:
Permutation speed up and minor fix.

austinhpatton committed:
Adds reconciliation-filtered gene family dating to the zoogle workflow, replaces in-process congruification with a dedicated treePL-based dating step using GeneRax speciation events, and includes a 12-species eukaryote test dataset for end-to-end debugging.

austinhpatton committed:
correctly handle hyphenated species names

austinhpatton committed:
removed batch handling of phylo-profiles - switched to per-family channels

austinhpatton committed:
fixed dated spp tree to val so all gene families are dated and phylo_dist runs

austinhpatton committed:
fix to ensure per-family phylo profile channels

austinhpatton committed:
fixed orthofinder phylohogs implementation

austinhpatton committed:
separate script to make correspondance for phylohogs

austinhpatton committed:
simplified reconciled event tree handling/output

austinhpatton committed:
removed legacy gft congruification output

austinhpatton committed:
simplified zoogle output path

austinhpatton committed:
phylo_dist renamed to zoogle

austinhpatton committed:
rename phylo_dist to zoogle

austinhpatton committed:
point modules to arcadiascience/zoogle:1.0.0 container

austinhpatton committed:
fix physicochemical_props COPY path to bin/zoogle

austinhpatton committed:
remove dead clinvar code

austinhpatton committed:
remove untracked files

austinhpatton committed:
add phangorn to versions.yml block

austinhpatton committed:
remove stale TODOs

austinhpatton committed:
standardized internal use of hypen Genus-species delimiter

austinhpatton committed:
apply bracket to species tree calibration, use fixed ages for gene trees.

austinhpatton committed:
updates to witch docker/singularity behavior

austinhpatton committed:
pared down euk test data

austinhpatton committed:
added container spec

austinhpatton committed:
fixed container spec and retry logic

austinhpatton committed:
rename first for consistency

austinhpatton committed:
updated defaults

austinhpatton committed:
removed OMA annotations

austinhpatton committed:
removed dead annotation Rscript

austinhpatton committed:
improved protein annotation strategy

austinhpatton committed:
updated schema

austinhpatton committed:
updated docs

austinhpatton committed:
updated changelog

austinhpatton committed:
Create Docker images and Makefile targets for preprocessing

austinhpatton committed:
Create Docker images and Makefile targets for preprocessing

austinhpatton committed:
build proteome preprocessing subworkflow and modules

austinhpatton committed:
updates to handle preprocessing/filtering

austinhpatton committed:
config and schema updates for preprocessing/filtering

austinhpatton committed:
samplesheet for proteome preprocessing and filtering

austinhpatton committed:
updated mcl inflation values for testing

austinhpatton committed:
updated process specs

austinhpatton committed:
separate download module

austinhpatton committed:
separate download module

austinhpatton committed:
fixed input data to transdecoder - nucleotides only

austinhpatton committed:
fixed check

austinhpatton committed:
folded og filtering into orthofinder module to streamline

austinhpatton committed:
flexibly allow local from paths, remote data from links, or ncbi reference genomes, or uniprot from  ncbi genome/uniprot proteome IDs

austinhpatton committed:
move busco and mcl testing to optional subworkflows

austinhpatton committed:
skip asteroid if no outgroup provided

austinhpatton committed:
sanitize sequence headers and preprend species prefix

austinhpatton committed:
fix tree handing

austinhpatton committed:
enforce gene family filters after alignment cleaning

austinhpatton committed:
lightweight implementation of merge phylo profiles

austinhpatton committed:
continued refactor of merge phylo profiles, streamlined output of profiles

austinhpatton committed:
keep only ungapped length >= 50

austinhpatton committed:
new three size-tiered alignment workflow: mafft -> witch -> famsa

austinhpatton committed:
depracate orthofinder phylohogs to use orthoxml for lightweight use of generax reconciliations

austinhpatton committed:
natively handle NCBI TSA transcriptome downloading and preprocessing

austinhpatton committed:
small fix

austinhpatton committed:
make mcl-test an optional field

austinhpatton committed:
removed brittle TSA handling

austinhpatton committed:
fixed isoform filtering and prequal memory allocation

austinhpatton committed:
fixed prequal memory handling

austinhpatton committed:
increase mcl cpus

austinhpatton committed:
dynamic process specifications

austinhpatton committed:
fixed rounding problem

austinhpatton committed:
fixed famsa specs

austinhpatton committed:
calculate physicochem properties on untrimmed sequences

austinhpatton committed:
updated test samplesheet

austinhpatton committed:
small genome TSA for test

austinhpatton committed:
correctly usin aa's in test

austinhpatton committed:
one more try

austinhpatton committed:
fix for physicochem properties module

austinhpatton committed:
fix to naming

austinhpatton committed:
fix to physicochem file writing

austinhpatton committed:
removed prequal, updated process specs

austinhpatton committed:
rescaled orthoxml memory

austinhpatton committed:
new, streamlined, lightweight phylohog parsing

austinhpatton committed:
updated to use storeDir rather than fragile cache for -resume

austinhpatton committed:
change container

austinhpatton committed:
syntax update for docker python version

austinhpatton committed:
fix to uniprot canonical proteome download

austinhpatton committed:
publishdir reorganization

austinhpatton committed:
add compositional entropy, per-protein physicochem/compositional entropy SD and autocorrelation to outputs

austinhpatton committed:
rename physicochem to protein properties

austinhpatton committed:
include protein properties module

austinhpatton committed:
remove unnecessary .first() invocations

austinhpatton committed:
rewired modules through focused subworkflows

austinhpatton committed:
simplify test runs with --test flag

austinhpatton committed:
simplify run specification with 'arcadia' config

austinhpatton committed:
renamed --test to --test_run

austinhpatton committed:
fixed missing .first() invocation

austinhpatton committed:
renamed orthofinder raw seq file directories from msas -> fas

austinhpatton committed:
more meaningful directory name for final processed proteomes

austinhpatton committed:
directory cleanup

austinhpatton committed:
evidence-based fasttree param specs

austinhpatton committed:
no retries for samplesheet check

austinhpatton committed:
non-optional outputs

austinhpatton committed:
remove unused --graphtraceoptimize flag in updated witch invocation

austinhpatton committed:
fixes to witch

austinhpatton committed:
updated witch memory allocation logic

austinhpatton committed:
increased mafft memory allocation

austinhpatton committed:
boosted mafft memory

austinhpatton committed:
boosted memory specs

austinhpatton committed:
boosted IQT mem specs

austinhpatton committed:
IQT -> 12 cpus

austinhpatton committed:
improved gene family tree time calibration retry logic

austinhpatton committed:
Rename MAFFT_ADAPTIVE -> MAFFT_TIER1 and consolidate adaptive_align param

austinhpatton committed:
Cap CPU requests at params.max_cpus across compute modules

austinhpatton committed:
Add versions.yml output to DATE_GENE_FAMILY_TREES and PARSE_PHYLOHOGS

austinhpatton committed:
Fix test samplesheet boolean values

austinhpatton committed:
Sync nextflow_schema.json with v2.0.0 defaults and new params

austinhpatton committed:
Update documentation for v2.0.0

austinhpatton committed:
updated memory and runtime

austinhpatton committed:
add centroid-based zoogle-analysis

austinhpatton committed:
updated zoogle docker image version

austinhpatton committed:
rewire with sentinal file to automatrically redirect failed MSA/trees to fallback processes

austinhpatton committed:
fix to handling of speciation calibration nodes

austinhpatton committed:
include paralog/ortholog/xenolog relationships to protein dist summary tables

austinhpatton committed:
time calibration edge case handling

austinhpatton committed:
handle small gene families

austinhpatton committed:
increase max_cpus for MCL clustering

austinhpatton committed:
switch simplified.config to use adaptive alignment mode

austinhpatton committed:
reverted to ignore on failure for MSA/tree fallbacks

austinhpatton committed:
documentation updates and streamlining

austinhpatton committed:
mermaid diagrams of workflow and subworkflows

austinhpatton committed:
remove reference to workflow figures, reference mermaid diagrams

austinhpatton committed:
remove devel doc

austinhpatton committed:
collapsible diagrams + input preparation diagram

austinhpatton committed:
minor reformatting of overview diagram

austinhpatton committed:
switch to using DL model for generax/speciesrax

austinhpatton committed:
improved uncertainty handling and calibration point distributions for gene family tree dating

austinhpatton committed:
improved documentation for samplesheet preparation

austinhpatton committed:
include calibration outputs in outputs.md

austinhpatton committed:
fix to perm tests for small families and improved storeDir checks for tree inference

austinhpatton committed:
Change to aws config for storeDir resume stability on large runs

austinhpatton committed:
resolve polytomies prior to running generax

austinhpatton committed:
resolve polytomies prior to running generax

austinhpatton committed:
same storeDir fix for MSAs as for trees

austinhpatton committed:
aws rate limiting fix on resuming partial run

austinhpatton committed:
update process specs and simplify gene family tree trime calibration

austinhpatton committed:
fix to timecalibration of gene family trees

austinhpatton committed:
fix to number sites calculation

austinhpatton committed:
update documentation and remove dead code

austinhpatton committed:
documentation updates and cleanup for release

austinhpatton committed:
fix iqtree singularity image specification

austinhpatton committed:
fix write-dir for timetree building

austinhpatton committed:
revamp samplesheet

austinhpatton committed:
clean up for release

austinhpatton committed:
explicit pointers to usage.md and output.md, and samplesheet descriptor

austinhpatton committed:
named version

austinhpatton committed:
update and cleanup for release

austinhpatton committed:
updated species tree gene family occupancy requirement
